### ISSUE TITLE
Glitches in generated stb_truetype bitmaps

### ISSUE THREAD

**Issue by blitz-research** [nothings/stb#1171]
Hi,

I implemented a little text renderer using stb_truetype the other day and initially it all looked pretty good, but then I got curious and started zooming in on the characters and noticed little 'notches' and 'dimples' in the bitmap.

So I knocked up an equivalent version in freetype and these glitches disappeared. Am I doing something wrong here, or is this just as good as it gets with stb_truetype?

Here's a screenshot of what I'm talking about. The image at the top is the freetype version, the one below is the stb_truetype one:

![stb_truetype_glitch](https://user-images.githubusercontent.com/3487554/125855210-598706ff-984d-4e72-a1e6-102df3ec4e59.png)

The font is 'Kanit-Regular.ttf' which I found at https://www.fontsquirrel.com but I've also tried 'bitstream vera sans mono' and get the same general effect.

I am using an 'glOrtho' style projection matrix at 160x90 in a 1280x720 glViewport sized window to 'zoom in' on the chars, and the font height is 64. Note that the stb_truetype font is quite a bit smaller than the freetype one. These settings do exaggerate the effect, it's not as bad at 'normal' size, but it's there and once you've seen it you can't un-see it!

It's possible my rendering code positioning is off by .5 here or there, but IMO this can't explain the glitches - it's still just drawing 'sprites' made up of 2 triangles per char, so the glitches must be in the generated bitmap.

My c++ font loading code is here (I've tried with both 'packed' and 'baked' versions, same result):

```cpp
#include "font.h"

//#define USE_FREETYPE

#ifdef USE_FREETYPE
#include <freetype2/ft2build.h>
#include FT_FREETYPE_H
#else
#define STB_RECT_PACK_IMPLEMENTATION
#include "stb_rect_pack.h"
#define STB_TRUETYPE_IMPLEMENTATION
#include "stb_truetype.h"
#endif

namespace wb {

namespace {
#ifdef USE_FREETYPE
FT_Library freetype;
#endif

} // namespace

float Font::textWidth(CString text) {

	float w = 0;
	for (size_t ch : text) {
		if (ch < firstChar || ch >= glyphs.size()) ch = firstChar;
		w += glyphs[ch - firstChar].advance;
	}
	return w;
}

#ifdef USE_FREETYPE

// freetype version...
Font* loadFont(CString path, float fontHeight, Device* device) {

	int firstChar = 32, numChars = 96, padding = 1;

	uint32_t atlasWidth = Font::atlasWidth;
	uint32_t atlasHeight = Font::atlasHeight;

	ensure(freetype || !FT_Init_FreeType(&freetype));

	auto fontData = loadData(path);
	ensure(fontData.size());

	FT_Face face;
	ensure(!FT_New_Memory_Face(freetype, fontData.data(), fontData.size(), 0, &face));

	ensure(!FT_Set_Pixel_Sizes(face, 0, fontHeight));
	float ascent = face->ascender;
	float descent = face->descender;
	float lineGap = face->height - (ascent + descent);

	Vector<uint8_t> atlasData(atlasWidth * atlasHeight);
	memset(atlasData.data(), 0, atlasData.size());

	Vector<Glyph> glyphs;

	Vec2f texScale = Vec2f(1.0f / atlasWidth, 1.0f / atlasHeight);
	int tx = padding, ty = padding, maxh = 0, baseline = 0;
	FT_GlyphSlot slot = face->glyph;

	for (int i = 0; i < numChars; ++i) {

		ensure(!FT_Load_Char(face, firstChar + i, FT_LOAD_RENDER));

		int w = slot->bitmap.width;
		int h = slot->bitmap.rows;

		if (tx + w + padding > atlasWidth) {
			tx = padding;
			ty += maxh + padding;
			maxh = 0;
		}
		maxh = std::max(maxh, h);

		uint8_t* src = slot->bitmap.buffer;
		uint8_t* dst = atlasData.data() + ty * atlasWidth + tx;
		for (int y = 0; y < h; ++y) {
			memcpy(dst, src, w);
			src += slot->bitmap.pitch;
			dst += atlasWidth;
		}

		int dx = slot->bitmap_left, dy = -slot->bitmap_top;

		//		debug() << "Glyph '" << char(i + firstChar) << "'" << w << h << dx << dy;

		baseline = std::max(baseline, -dy);

		Rectf srcRect(Vec2f(tx, ty) * texScale, Vec2f(tx + w, ty + h) * texScale);
		Rectf dstRect(Vec2f(dx, dy), Vec2f(dx + w, dy + h));

		glyphs.emplace_back(srcRect, dstRect, float(slot->advance.x) / 64);

		tx += w + padding;
	}

	FT_Done_Face(face);

	Texture* atlas = device->createTexture(atlasData.data(), atlasWidth, atlasHeight, TextureFormat::alpha8);

	return new Font(atlas, std::move(glyphs), firstChar, fontHeight, ascent, descent, lineGap, baseline);
}

#else

// stb_truetype version...
Font* loadFont(CString path, float fontHeight, Device* device) {

	int firstChar = 32, numChars = 96;

	uint32_t atlasWidth = Font::atlasWidth;
	uint32_t atlasHeight = Font::atlasHeight;

	auto fontData = loadData(path);
	ensure(fontData.size());

	stbtt_fontinfo info;
	ensure(stbtt_InitFont(&info, fontData.data(), 0));

	int ascent, descent, lineGap;
	stbtt_GetFontVMetrics(&info, &ascent, &descent, &lineGap);
	float fscale = fontHeight / (ascent - descent);

	Vector<uint8_t> atlasData(atlasWidth * atlasHeight);

	Vector<Glyph> glyphs;

#ifdef STB_RECT_PACK_IMPLEMENTATION

	stbtt_pack_context packer;
	ensure(stbtt_PackBegin(&packer, atlasData.data(), atlasWidth, atlasHeight, 0, 1, nullptr));

	stbtt_PackSetOversampling(&packer, 1, 1);

	Vector<stbtt_packedchar> charData(numChars);

	stbtt_PackFontRange(&packer, fontData.data(), 0, fontHeight, firstChar, numChars, charData.data());

	stbtt_PackEnd(&packer);

	float baseline = 0;
	Vec2f texScale = Vec2f(1.0f / atlasWidth, 1.0f / atlasHeight);

	for (size_t i = 0; i < 96; ++i) {

		const auto& ch = charData[i];

		baseline = std::max(baseline, -ch.yoff);

		Rectf srcRect = {Vec2f(ch.x0, ch.y0) * texScale, Vec2f(ch.x1, ch.y1) * texScale};
		Rectf dstRect = {Vec2f(ch.xoff, ch.yoff), Vec2f(ch.xoff2, ch.yoff2)};

		glyphs.emplace_back(srcRect, dstRect, ch.xadvance);
	}

#else

	Vector<stbtt_bakedchar> charData(numChars);

	stbtt_BakeFontBitmap(fontData.data(), 0, fontHeight, atlasData.data(), atlasWidth, atlasHeight, firstChar, numChars,
						 charData.data());

	float baseline = 0;
	Vec2f texScale = Vec2f(1.0f / atlasWidth, 1.0f / atlasHeight);

	for (size_t i = 0; i < 96; ++i) {

		const auto& ch = charData[i];

		baseline = std::max(baseline, -ch.yoff);

		Rectf srcRect = {Vec2f(ch.x0, ch.y0) * texScale, Vec2f(ch.x1, ch.y1) * texScale};
		Rectf dstRect = {Vec2f(ch.xoff, ch.yoff), Vec2f(ch.xoff + ch.x1 - ch.x0, ch.yoff + ch.y1 - ch.y0)};

		glyphs.emplace_back(srcRect, dstRect, ch.xadvance);
	}

#endif

	Texture* atlas = device->createTexture(atlasData.data(), atlasWidth, atlasHeight, TextureFormat::alpha8);

	return new Font(atlas, std::move(glyphs), firstChar, fontHeight, ascent * fscale, descent * fscale,
					lineGap * fscale, baseline);
}

#endif

} // namespace wb
```

**Comment by stephomi** [comment]
I got the same issue, it happens with any fonts.

It seems to be caused by https://github.com/nothings/stb/commit/0be82e4814eaad3b9af6261ee1355608a0d04b0e#diff-fe475a15d358f0069f6344900123a07ebfa13460a2340122d51e0421adc29202

**Comment by nothings** [comment]
The change in question was made because there was an assert being tripped even though there was no visual failure, and I looked at the code and realized it was wrong and attempted to make it right, and didn't notice any flaws in when I tested. As a temp workaround, do this:

`#define STBTT_RASTERIZER_VERSION 1`

before including, which will be slower but should be correct.

**Comment by Feacur** [comment]
ah, didn't notice it initially, so I've created a new issue: https://github.com/nothings/stb/issues/1189

**Comment by nothings** [comment]
fixed in 1.26, available now
